### ISSUE TITLE
Policy: Recurrent LSTM policy for partial observability

### ISSUE THREAD

**Issue by espenhk** [espenhk/gamer-ai#22]
## Summary

Add an **LSTM-based recurrent policy** (`LSTMPolicy`) that maintains a hidden state across steps within an episode. This addresses the fundamental partial observability limitation of all current policies: a single-frame observation cannot distinguish states that look identical but require different actions (e.g. entering a blind corner at high vs low speed after a jump).

## Motivation

All current policies in `policies.py` are **memoryless Markov** — they map `obs → action` without any recollection of prior steps. This is a known limitation for racing games where:

1. **Momentum**: the car's recent velocity history predicts upcoming slides better than instantaneous `angular_vel`
2. **Track sections**: parts of `a03` beyond the lookahead window require knowing which segment you came from
3. **Input lag**: steering changes take multiple ticks to produce observable lateral response; memory helps model this
4. **Repeated structures**: two geometrically similar track sections may need different lines due to different approach speeds

The current 3-point lookahead (`LOOKAHEAD_STEPS = [10, 25, 50]`) partially compensates for this, but LSTM can learn arbitrary temporal patterns from training data.

## Proposed Implementation

```python
class LSTMPolicy(BasePolicy):
    """
    Single-layer LSTM policy (pure numpy).
    Hidden state (h, c) persists across steps within an episode.
    Trained via REINFORCE or CMA-ES over episode returns.
    """
```

### Architecture

```
Input at each step:  obs  (21 + n_lidar_rays dims)
LSTM layer:          [obs; h_prev] → h_t, c_t  (hidden_size units)
Output heads:
  steer:   tanh(W_s @ h_t)            → [-1, 1]
  accel:   sigmoid(W_a @ h_t) > 0.5   → {0, 1}
  brake:   sigmoid(W_b @ h_t) > 0.5   → {0, 1}
```

LSTM equations (standard):
```
f = σ(W_f @ [h; x] + b_f)   # forget gate
i = σ(W_i @ [h; x] + b_i)   # input gate
g = tanh(W_g @ [h; x] + b_g) # cell gate
o = σ(W_o @ [h; x] + b_o)   # output gate
c = f * c_prev + i * g
h = o * tanh(c)
```

### Training strategy

`LSTMPolicy` uses **CMA-ES or GeneticPolicy** as its outer optimiser — the full parameter vector (all LSTM + output head weights flattened) is treated as the genotype. This avoids implementing BPTT and keeps the training loop identical to existing evolutionary methods.

```
flat_params = [W_f, b_f, W_i, b_i, W_g, b_g, W_o, b_o, W_s, W_a, W_b, ...]
```

Forward pass is handled by `LSTMPolicy.__call__`; the outer optimiser mutates `flat_params` exactly as it does for `WeightedLinearPolicy`.

### Episode lifecycle

```python
def on_episode_end(self):
    self._h = np.zeros(hidden_size)   # reset hidden state
    self._c = np.zeros(hidden_size)   # reset cell state

def __call__(self, obs):
    self._h, self._c = lstm_forward(obs, self._h, self._c)
    steer = tanh(W_s @ self._h)
    accel = float(sigmoid(W_a @ self._h) > 0.5)
    brake  = float(sigmoid(W_b @ self._h) > 0.5)
    return np.array([steer, accel, brake])
```

### Config additions

```yaml
policy_type: lstm
policy_params:
  hidden_size: 32
  outer_optimizer: cmaes   # or "genetic"
  population_size: 20
  initial_sigma: 0.05
```

## Files to change

| File | Change |
|------|--------|
| `policies.py` | Add `LSTMPolicy` + `_lstm_forward()` helper (~120 lines) |
| `main.py` `_make_policy()` | Register `"lstm"` case; delegate outer optimiser to CMA-ES or GeneticPolicy |
| `config/training_params.yaml` | Document policy + params |
| `tests/test_policies.py` | Unit test: hidden state updates correctly; episode reset clears state |

## Key Design Decisions

- **Pure numpy LSTM**: consistent with existing codebase; no torch dependency.
- **Evolution as outer optimiser**: sidesteps BPTT through time, which requires differentiable reward signal. CMA-ES handles any black-box fitness function.
- **Hidden size 32**: small enough that the flat parameter vector (~4×(32+21)×32 ≈ 5,504 params) is manageable for CMA-ES/genetic with population ~20.
- **Episode reset hook**: `on_episode_end()` clears `(h, c)` — consistent with existing hook pattern.
- **YAML save**: store LSTM weight matrices as nested lists (same schema as `NeuralNetPolicy`) + hidden_size.

## Acceptance Criteria

- [ ] `LSTMPolicy.__call__` updates `(h, c)` in-place each step and returns correct action shape
- [ ] `on_episode_end()` resets hidden state to zeros
- [ ] Flat parameter vector round-trips through `to_flat()` / `with_flat()` without shape errors
- [ ] Works as a drop-in with `GeneticPolicy` outer optimiser (swap inner individual type)
- [ ] `policy_type: lstm` works end-to-end in `main.py`
- [ ] Unit test: two identical obs sequences with different prior history produce different actions

## References

- Hochreiter & Schmidhuber (1997). *Long Short-Term Memory.* Neural Computation 9(8).
- Salimans et al. (2017). *Evolution Strategies as a Scalable Alternative to Reinforcement Learning.* arXiv:1703.03864
